test/utils/shoots/update/highavailability/update.go

```diff
@@ -0,0 +1,222 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package highavailability
+
+import (
+	"context"
+
+	gardencorev1beta1 "github.com/gardener/gardener/pkg/apis/core/v1beta1"
+	gardencorev1beta1helper "github.com/gardener/gardener/pkg/apis/core/v1beta1/helper"
+
+	"github.com/gardener/gardener/pkg/client/kubernetes"
+	"github.com/gardener/gardener/test/framework"
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+	appsv1 "k8s.io/api/apps/v1"
+	batchv1 "k8s.io/api/batch/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/utils/pointer"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// UpdateAndVerify runs the HA control-plane update tests for an existing shoot cluster.
+func UpdateAndVerify(
+	ctx context.Context,
+	f *framework.ShootFramework,
+	failureToleranceType gardencorev1beta1.FailureToleranceType,
+
+) {
+	By("Update Shoot control plane to HA with failure tolerance type " + string(failureToleranceType))
+	Expect(f.UpdateShoot(ctx, func(shoot *gardencorev1beta1.Shoot) error {
+		shoot.Spec.ControlPlane = &gardencorev1beta1.ControlPlane{
+			HighAvailability: &gardencorev1beta1.HighAvailability{
+				FailureTolerance: gardencorev1beta1.FailureTolerance{
+					Type: failureToleranceType,
+				},
+			},
+		}
+		return nil
+	})).To(Succeed())
+
+	By("Verify Shoot's control plane components")
+	verifyTSC(ctx, f.SeedClient, f.Shoot, f.ShootSeedNamespace())
+	verifyEtcdAffinity(ctx, f.SeedClient, f.Shoot, f.ShootSeedNamespace())
+}
+
+func getTSCForZone(labels map[string]string) corev1.TopologySpreadConstraint {
+	return corev1.TopologySpreadConstraint{
+		TopologyKey:       corev1.LabelTopologyZone,
+		MaxSkew:           1,
+		WhenUnsatisfiable: corev1.DoNotSchedule,
+		LabelSelector:     &metav1.LabelSelector{MatchLabels: labels},
+	}
+}
+
+func getTSCForNode(labels map[string]string) corev1.TopologySpreadConstraint {
+	return corev1.TopologySpreadConstraint{
+		TopologyKey:       corev1.LabelHostname,
+		MaxSkew:           1,
+		WhenUnsatisfiable: corev1.DoNotSchedule,
+		LabelSelector:     &metav1.LabelSelector{MatchLabels: labels},
+	}
+}
+
+func verifyTSC(ctx context.Context, seedClient kubernetes.Interface, shoot *gardencorev1beta1.Shoot, namespace string) {
+	c := seedClient.Client()
+	commponents := map[string]map[string]string{
+		"gardener-resource-manager": {
+			"app":                 "gardener-resource-manager",
+			"gardener.cloud/role": "controlplane",
+		},
+		"kube-apiserver": {
+			"app":  "kubernetes",
+			"role": "apiserver",
+		},
+	}
+
+	for name, l := range commponents {
+		deployment := &appsv1.Deployment{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      name,
+				Namespace: namespace,
+			},
+		}
+		Expect(c.Get(ctx, client.ObjectKeyFromObject(deployment), deployment)).To(Succeed())
```

Comment:
From https://github.com/gardener/gardener/blob/master/docs/development/testing.md#writing-tests-relevant-for-all-kinds:

> make sure to add additional descriptions to Gomega matchers if necessary (e.g. in a loop): [example test](https://github.com/gardener/gardener/blob/2eb54485231408cbdbabaa49812572a07124364f/test/e2e/shoot/internal/rotation/certificate_authorities.go#L89-L93)

Please follow this.